### PR TITLE
[ONNX] ONNX Exporter logging

### PR DESCRIPTION
**Pull request by BowenBao**
Add ONNX exporter logging facility. Supporting both C++/Python logging api. Logging can be turned on/off. Logging output stream can be either set to `stdout` or `stderr`.

A few other changes:
* When exception is raised in passes, the current IR graph being processed will be logged.
* When exception is raised from `_jit_pass_onnx` (the pass that converts nodes from namespace `ATen` to `ONNX`), both ATen IR graph and ONNX IR graph under construction will be logged.
* Exception message for ConstantFolding is truncated to avoid being too verbose.
* Update the final printed IR graph with node name in ONNX ModelProto as node attribute. Torch IR Node does not have name. Adding this to printed IR graph helps debugging.

### CODE REVIEW

File: torch/csrc/jit/python/init.cpp [hunk 1]
```diff
@@ -292,6 +293,37 @@ void initJITBindings(PyObject* module) {
...
+          [](std::string stream_name) -> void {
+            if (stream_name == "stdout") {
+              ::torch::jit::set_onnx_logging_output_stream(std::cout);
+            } else if (stream_name == "stderr") {
+              ::torch::jit::set_onnx_logging_output_stream(std::cerr);
+            } else {
+              std::cerr << "ERROR: only `stdout` and `stderr`"
+                        << "are supported as output options" << std::endl;
+            }
+          })
```
Comment (thiagocrepaldi) [inline torch/csrc/jit/python/init.cpp:315]: Have you considered merging this code into `jit_init`? There can be a default value (disabled), but the user could enable it and pass the desired stream during jit initialization.

This way there is no need to add more APIs

File: torch/csrc/jit/passes/onnx/onnx_log.h [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+namespace torch {
+namespace jit {
+
+TORCH_API bool get_onnx_logging_enabled();
+
+TORCH_API void set_onnx_logging_enabled(bool enabled);
+
+TORCH_API void set_onnx_logging_output_stream(std::ostream& out_stream);
+
+TORCH_API std::ostream& get_onnx_logging_output_stream();
```
Comment (thiagocrepaldi) [inline torch/csrc/jit/passes/onnx/onnx_log.h:15]:
```suggestion
TORCH_API std::ostream& get_onnx_logging_handler();
```

Maybe a rename here to dont leak implementation detail (output stream is an implementation detail)

File: torch/csrc/jit/passes/onnx/onnx_log.h [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+#include <string>
+
+namespace torch {
+namespace jit {
+
+TORCH_API bool get_onnx_logging_enabled();
+
+TORCH_API void set_onnx_logging_enabled(bool enabled);
+
+TORCH_API void set_onnx_logging_output_stream(std::ostream& out_stream);
```
Comment (thiagocrepaldi) [inline torch/csrc/jit/passes/onnx/onnx_log.h:13]: Maybe that could be done during `jit_init`, saving a couple more APIs

**Review by thiagocrepaldi — COMMENTED** [review]
Overall looks great. Just have made a comment on how to initialize/enabled/disable the log

File: torch/csrc/jit/python/init.cpp (hunk 1 above)
Comment (BowenBao) [reply torch/csrc/jit/python/init.cpp]: merged the two

File: torch/csrc/jit/passes/onnx/onnx_log.h (hunk 2 above)
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/onnx_log.h]: renamed with underscore to mark as internal api. I'll keep it simple for now naming with the return class.

File: torch/onnx/__init__.py [hunk 4]
```diff
@@ -412,3 +412,30 @@ def unregister_custom_op_symbolic(symbolic_name, opset_version):
...
+      stream_name (str, default "stdout"): Only ``stdout`` and ``stderr`` are supported
+        as output options.
+    """
+    _C._jit_set_onnx_logging_enabled(enabled, stream_name)
+
+
+def log(*args):
+    r"""
+    A simple logging facility for ONNX exporter.
+    """
```
[the remark this answers is not in the document]
Comment (thiagocrepaldi) [reply torch/onnx/__init__.py:440]: Looking `torch.onnx.log` calls below, it seems users are responsible for adding `\n` at the end of the line. If that is the case, that should be specified in the docstring. We should handle the difference across platforms `\n` vs `\r\n` vs `\r`

File: docs/source/onnx.rst [hunk 5]
```diff
@@ -598,6 +598,9 @@ Functions
 .. autofunction:: register_custom_op_symbolic
 .. autofunction:: select_model_mode_for_export
 .. autofunction:: is_in_onnx_export
+.. autofunction:: onnx_logging_enabled
```
Comment (thiagocrepaldi) [inline docs/source/onnx.rst:601]:
```suggestion
.. autofunction:: is_onnx_log_enabled
```
All existing APIs above start with a verb.

File: docs/source/onnx.rst [hunk 6]
```diff
@@ -598,6 +598,9 @@ Functions
...
+.. autofunction:: set_onnx_logging_enabled
```
Comment (thiagocrepaldi) [inline docs/source/onnx.rst:602]:
```suggestion
.. autofunction:: config_log
```
This setter does more than just enabling/disabling the log, so a more comprehensive name would be more intuitive. 

`(Python's logging module)[https://docs.python.org/3/library/logging.html#logging.basicConfig]` exposes the method `basicConfig`. Maybe we could do something similar such as `torch.onnx.config_log` to specify both status and/or stream.

Having `torch.onnx.enable_log()`, `torch.onnx.disable_log()` and `torch.onnx.set_stream(stream_name: str)` could be also intuitive and easy to use

File: torch/csrc/jit/passes/onnx/onnx_log.cpp [hunk 7]
```diff
@@ -0,0 +1,26 @@
+#include <torch/csrc/jit/passes/onnx/onnx_log.h>
+#include <iostream>
+
+namespace torch {
+namespace jit {
+
+static bool onnx_logging_enabled = false;
+static std::ostream* out = &std::cout;
+
+bool get_onnx_logging_enabled() {
```
Comment (thiagocrepaldi) [inline torch/csrc/jit/passes/onnx/onnx_log.cpp:10]:
```suggestion
bool is_onnx_logging_enabled() {
```

File: docs/source/onnx.rst (hunk 6 above)
Comment (BowenBao) [reply docs/source/onnx.rst:602]: Thank you for suggestions! I also like the second suggestion, which is very clean. I want to keep it simple that way.

File: torch/onnx/__init__.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/onnx/__init__.py:440]: The `\n` is in the middle, users don't need to explicitly add `\n` for the end of the line. I'll remove it as it appears misleading already.

**Review by thiagocrepaldi — APPROVED** [review]

**Comment by thiagocrepaldi** [comment]
@BowenBao Will this be merged into pytorch:master directly or should we change it to pytorch:onnx_ms_1?

**Comment by BowenBao** [comment]
> @BowenBao Will this be merged into pytorch:master directly or should we change it to pytorch:onnx_ms_1?

I'm planning to directly merge into master.

File: torch/csrc/jit/passes/onnx/onnx_log.cpp [hunk 8]
```diff
@@ -0,0 +1,29 @@
+#include <torch/csrc/jit/passes/onnx/onnx_log.h>
+#include <iostream>
+
+namespace torch {
+namespace jit {
+
+static bool onnx_log_enabled = false;
+static std::ostream* out = &std::cout;
```
Comment (malfet) [inline torch/csrc/jit/passes/onnx/onnx_log.cpp:8]:
```suggestion
static std::shared_ptr<std::ostream> out;
```

File: torch/csrc/jit/passes/onnx/onnx_log.cpp [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+void set_onnx_log_enabled(bool enabled) {
+  onnx_log_enabled = enabled;
+}
+
+void set_onnx_log_output_stream(std::ostream* out_stream) {
+  if (nullptr != out_stream) {
+    out = out_stream;
+  }
+}
```
Comment (malfet) [inline torch/csrc/jit/passes/onnx/onnx_log.cpp:22]: Seen raw pointers in C++ code in 2022 is a bit weird. Who guarantees the lifetime of the object?
Let's use `shared_ptr` to ensure that out is valid unless somebody resets it
```suggestion
void set_onnx_log_output_stream(std::shared_ptr<std::ostream> out_stream) {
    out = std::move(out_stream);
}
```

File: torch/csrc/jit/passes/onnx/onnx_log.cpp [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+std::ostream& _get_onnx_log_output_stream() {
+  return *out;
```
Comment (malfet) [inline torch/csrc/jit/passes/onnx/onnx_log.cpp:25]:
```suggestion
  return out? *out : std::cout;
```

File: torch/csrc/jit/passes/onnx/onnx_log.cpp (hunk 8 above)
Comment (malfet) [inline torch/csrc/jit/passes/onnx/onnx_log.cpp:8]: Anonymous namespace vs static?

File: torch/csrc/jit/passes/onnx/onnx_log.cpp [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace torch {
+namespace jit {
+
+static bool onnx_log_enabled = false;
+static std::ostream* out = &std::cout;
+
+bool is_onnx_log_enabled() {
+  return onnx_log_enabled;
+}
```
Comment (malfet) [inline torch/csrc/jit/passes/onnx/onnx_log.cpp:12]: Wouldn't it be better to declare all those function in `onnx` namespace, and simply call them `onnx::is_log_enabled()` etc?

File: torch/csrc/jit/passes/onnx/onnx_log.h [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+TORCH_API bool is_onnx_log_enabled();
+
+TORCH_API void set_onnx_log_enabled(bool enabled);
+
+TORCH_API void set_onnx_log_output_stream(std::ostream* out_stream);
+
+TORCH_API std::ostream& _get_onnx_log_output_stream();
+
+#define ONNX_LOG(...)                            \
+  if (is_onnx_log_enabled()) {                   \
```
Comment (malfet) [inline torch/csrc/jit/passes/onnx/onnx_log.h:18]: Defines are not guaranteed to be invoked in the same namespace, so please specify namespace fully (same as you do on line 19) 
```suggestion
  if (::torch::jit::is_onnx_log_enabled()) {                   \
```

**Comment by thiagocrepaldi** [comment]
@BowenBao what is procedure when a PR waiting to be imported conflicts with master? do we have to rebase it and ask for another import?

**Comment by BowenBao** [comment]
> @BowenBao what is procedure when a PR waiting to be imported conflicts with master? do we have to rebase it and ask for another import?

If the PR is not imported, you should rebase it.
If it was already imported, it's better to check with Meta employee if a rebase and reimport is needed.

**Comment by malfet** [comment]
Trying to land again

**Pull request closed without merging**